### ISSUE TITLE
Add flag to allow fork-ts-checker-webpack-plugin to check for syntax errors (HappyPack / thread-loader)

### ISSUE THREAD

**Issue by johnnyreilly** [TypeStrong/fork-ts-checker-webpack-plugin#39]
Excuse the minimal post; on my phone. An issue was logged against ts-loader; details can be found here:

https://github.com/TypeStrong/ts-loader/issues/602

Essentially this comes down to ts-loader being unable to report errors to webpack when used in `happyPackMode` ie ts-loader is used with HappyPack or thread-loader to split work over multiple instances

**Comment by johnnyreilly** [comment]
copied from linked issue in ts-loader:

I think the issue is here: https://github.com/TypeStrong/ts-loader/blob/master/src/index.ts#L214

```ts
    // _module.errors is not available inside happypack - see https://github.com/TypeStrong/ts-loader/issues/336
    if (!instance.loaderOptions.happyPackMode) {
        registerWebpackErrors(
            loader._module.errors,
            formatErrors(diagnostics, instance.loaderOptions, instance.compiler, { module: loader._module })
        );
    }
```

In `happyPackMode` we cannot register transpile errors with webpack and so this info is lost. BTW the same issue would presumably apply if using thread-loader instead of happypack.  Same principle; ts-loader cannot register errors with webpack.

There's nothing that we can do from a ts-loader perspective here.  

>  I think it should not be hard to add flag to fork-ts-checker plugin to check syntax errors.

If you could look at this that would be awesome!

**Comment by abirmingham** [comment]
Currently we are seeing an issue where a number of errors are not reported by ts-loader and are not reported by fork-ts-checker-webpack-plugin. This is fairly concerning :)

Examples include:
```js
// Note the missing comma below:
const bar = [
    'a',
    'b'
    'c'
];

// Note the semi-colons below:
const biz = {
    a: 1,
    b: 2;
    c: 3;
    d: 4
};

// Note the missing '=>' below:
export const foo = async (bar:number):Promise<void> {
   ...
}
```

@johnnyreilly would this issue cause the behavior we're seeing? Thanks!

**Comment by johnnyreilly** [comment]
I think so yes @abirmingham. If you drop usage of thread-loader / cache-loader and stop using happypack mode for the time being the issues will be raised as you'd hope.

In the ts-loader issue @piotr-oles said:

> fork-ts-checker reports only semantic errors, syntax errors should be checked by ts-loader and as far as I know it's how it works in transpileOnly mode. Maybe it's because of happy-pack mode? I think it should not be hard to add flag to fork-ts-checker plugin to check syntax errors.

And so I'm hoping he's up for adding this flag. :smile: @piotr-oles would you be able to advise if you're likely to do this anytime soon? Failing that would you be able to advise on how you had in mind to do that so someone else could have a go and submit a PR? Thanks! 

Oh and thanks for all your work on this plugin; it's really useful. 💓

**Comment by johnnyreilly** [comment]
Having done a little digging I've an idea for a fix.  I think that the tweak needs to be made here: https://github.com/Realytics/fork-ts-checker-webpack-plugin/blob/master/lib/IncrementalChecker.js#L128

I think that this line:

```js
diagnostics.push.apply(diagnostics, this.program.getSemanticDiagnostics(sourceFile, cancellationToken));
```

Should become something like this:

```js
diagnostics.push.apply(diagnostics, this.program.getSemanticDiagnostics(sourceFile, cancellationToken), getSyntacticErrors ? this.program.getSyntacticDiagnostics(sourceFile, cancellationToken) : []);
```

`getSyntacticErrors` would need to be an option that was supplied.

**Comment by johnnyreilly** [comment]
Hi @abirmingham,

I've implemented a fix.  Do you want to give my fork a try?  You can find it here: https://github.com/johnnyreilly/fork-ts-checker-webpack-plugin

You'll need to supply a new option called `checkSyntacticErrors` to ensure that syntactic errors are reported, eg:

```
        new ForkTsCheckerWebpackPlugin({ checkSyntacticErrors: true })
```

It worked on my local projects just fine.  I've submitted a PR - hopefully it will get merged soon.  If you could confirm the PR resolves your issues that would be super helpful.

**Comment by johnnyreilly** [comment]
cc @voodoorider

**Comment by voodoorider** [comment]
@johnnyreilly It worked, however after linking your fork in `thread-loader` example I discovered that the syntactical errors are reported even without providing the `checkSyntacticErrors` flag.

**Comment by johnnyreilly** [comment]
@voodoorider - thanks for the feedback. I think that means that this change possibly **always** turns on `checkSyntacticErrors` flag.  This would explain the strange behaviour I experienced with the tests.  I think I may need to look at how the option is passed through....

cc @piotr-oles

**Comment by johnnyreilly** [comment]
I think I might have fixed it @voodoorider  - want to try now?

**Comment by johnnyreilly** [comment]
Closed by #43

**Comment by wowczarczyk** [comment]
@johnnyreilly @piotr-oles works perfectly, thanks guys!

**Comment by johnnyreilly** [comment]
Great!
